src/lpython/semantics/python_comptime_eval.h

```diff
@@ -242,6 +247,107 @@ struct PythonIntrinsicProcedures {
+        }
+        ASR::expr_t* arg1 = args[0], *arg2 = args[1];
+        LFORTRAN_ASSERT(ASRUtils::check_equal_type(ASRUtils::expr_type(arg1),
+                                    ASRUtils::expr_type(arg2)));
+        ASR::ttype_t* type = ASRUtils::expr_type(arg1);
+        if (ASRUtils::is_integer(*type)) {
+            int64_t a = ASR::down_cast<ASR::ConstantInteger_t>(arg1)->m_n;
+            int64_t b = ASR::down_cast<ASR::ConstantInteger_t>(arg2)->m_n;
+            return ASR::down_cast<ASR::expr_t>(
+                ASR::make_ConstantInteger_t(al, loc, (a^b), type));
+        } else {
+            throw SemanticError("_bitwise_xor() must have both integer arguments.", loc);
+        }
+    }
+
+    static ASR::expr_t *eval__bitwise_lshift(Allocator &al, const Location &loc, Vec<ASR::expr_t*> &args) {
+        LFORTRAN_ASSERT(ASRUtils::all_args_evaluated(args));
+        if (args.size() != 2) {
+            throw SemanticError("_bitwise_lshift() must have two integer arguments.", loc);
+        }
+        ASR::expr_t* arg1 = args[0], *arg2 = args[1];
+        LFORTRAN_ASSERT(ASRUtils::check_equal_type(ASRUtils::expr_type(arg1),
+                                    ASRUtils::expr_type(arg2)));
+        ASR::ttype_t* type = ASRUtils::expr_type(arg1);
+        if (ASRUtils::is_integer(*type)) {
+            int64_t a = ASR::down_cast<ASR::ConstantInteger_t>(arg1)->m_n;
+            int64_t b = ASR::down_cast<ASR::ConstantInteger_t>(arg2)->m_n;
+            if (b < 0) {
+                throw SemanticError("Negative shift count not allowed.", loc);
+            }
+            return ASR::down_cast<ASR::expr_t>(
+                ASR::make_ConstantInteger_t(al, loc, (a<<b), type));
+        } else {
+            throw SemanticError("_bitwise_lshift() must have both integer arguments.", loc);
+        }
+    }
+
+    static ASR::expr_t *eval__bitwise_rshift(Allocator & al, const Location &loc, Vec<ASR::expr_t*> &args) {
+        LFORTRAN_ASSERT(ASRUtils::all_args_evaluated(args));
+        if (args.size() != 2) {
+            throw SemanticError("_bitwise_rshift() must have two integer arguments.", loc);
+        }
+        ASR::expr_t* arg1 = args[0], *arg2 = args[1];
+        LFORTRAN_ASSERT(ASRUtils::check_equal_type(ASRUtils::expr_type(arg1),
+                                    ASRUtils::expr_type(arg2)));
+        ASR::ttype_t* type = ASRUtils::expr_type(arg1);
+        if (ASRUtils::is_integer(*type)) {
+            int64_t a = ASR::down_cast<ASR::ConstantInteger_t>(arg1)->m_n;
+            int64_t b = ASR::down_cast<ASR::ConstantInteger_t>(arg2)->m_n;
+            if (b < 0) {
+                throw SemanticError("Negative shift count not allowed.", loc);
+            }
+            return ASR::down_cast<ASR::expr_t>(
+                ASR::make_ConstantInteger_t(al, loc, (a>>b), type));
+        } else {
+            throw SemanticError("_bitwise_rshift() must have both integer arguments.", loc);
+        }
+    }
+
```

Comment:
Since the operations or process repeat the same in each function,  we can replace it with marcos, something like this:
```suggestion
#define BITWISE(X) \
    static ASR::expr_t *X(Allocator &al, const Location &loc, Vec<ASR::expr_t*> &args) { \
        return eval_bitwise(al, loc, args, &X); \
    }

BITWISE(eval__bitwise_or)
BITWISE(eval__bitwise_and)
BITWISE(eval__bitwise_xor)
BITWISE(eval__bitwise_lshift)
BITWISE(eval__bitwise_rshift)

    static ASR::expr_t *eval_bitwise(Allocator &al, const Location &loc,
                Vec<ASR::expr_t*> &args, comptime_eval_callback X) {
        LFORTRAN_ASSERT(ASRUtils::all_args_evaluated(args));
        if (args.size() != 2) {
            throw SemanticError("Bitwise Operation must have two integer arguments.", loc);
        }
        ASR::expr_t* arg1 = args[0], *arg2 = args[1];
        LFORTRAN_ASSERT(ASRUtils::check_equal_type(ASRUtils::expr_type(arg1),
                                    ASRUtils::expr_type(arg2)));
        ASR::ttype_t* type = ASRUtils::expr_type(arg1);
        if (ASRUtils::is_integer(*type)) {
            int64_t a = ASR::down_cast<ASR::ConstantInteger_t>(arg1)->m_n;
            int64_t b = ASR::down_cast<ASR::ConstantInteger_t>(arg2)->m_n;
            int64_t result;
            if(X == eval__bitwise_or) {
                result = a | b;
            } else if(X == eval__bitwise_and) {
                result = a & b;
            } else if(X == eval__bitwise_xor) {
                result = a ^ b;
            } else if(X == eval__bitwise_lshift) {
                result = a << b;
            } else if(X == eval__bitwise_rshift) {
                result = a >> b;
            }
            return ASR::down_cast<ASR::expr_t>(ASR::make_ConstantInteger_t(al, loc, result, type));;
        } else {
            throw SemanticError("Bitwise Operation must have both integer arguments.", loc);
        }
    }

```

Reply:
Excellent, thanks!

Reply:
Perfect, this looks good now.